ext/tbb/tbb_spin_rw_mutex.h

```diff
@@ -0,0 +1,291 @@
+ *      'read_release' calls and then it will do one 'write_acquire' and
+ *      'write_release'. And then if more work to be done, repeat.
+ *
+ *      For a thread:
+ *
+ *      num readers
+ *      ----------- = 2^(log2_ratio) - 1;
+ *      num writers
+ *
+ *       > log2_ratio of  0 means all writers
+ *       > log2_ratio of ~0 means all readers
+ *       > default log2_ratio is 6 e.g 63 reads per write.
+ *
+ *      - Pure readers are CPUs which will never perform a write acq/rel. The
+ *        cmdline arg is a bit mask e.g. 0x8 will make 4th cpu (cpu id: 0x3)
+ *        a pure reader. Default is 0x0 e.g. no pure readers.
+ *
+ */
+
+#ifndef __TBB_spin_mutex_H
+#define __TBB_spin_mutex_H
+
+#define initialize_lock(lock, threads) tbb_init_locks(lock, threads)
+#define parse_test_args(args, argc, argv) tbb_parse_args(args, argc, argv)
+
+#include "tbb.h"
+
+#define WRITER          1
+#define WRITER_PENDING  2
+#define READERS         ~(WRITER | WRITER_PENDING)
+#define ONE_READER      4
+#define BUSY            (WRITER | READERS)
+
+unsigned long log2_ratio = 0;
+unsigned long rw_mask = 0;
+unsigned long reader_cpu_mask = 0;
+
+typedef struct {
+    unsigned long c;
+    uint8_t pure_reader;
+} __attribute__((aligned(64))) rw_count_t;
+
+rw_count_t *rw_counts;
+
+inline uint8_t is_writer(unsigned long i, uint8_t val) {
+    if (rw_counts[i].pure_reader)
+        return 0;
+    rw_counts[i].c += val;
+    return !(rw_counts[i].c & rw_mask);
+}
+
+void tbb_print_usage() {
+    fprintf(stderr, " tbb_spin_rw_mutex\n");
+    fprintf(stderr, "\t[-h print this msg]\n");
+    fprintf(stderr, "\t[-r reader/writer log ratio, default: 6 (2^(6)-1 readers per writer)]\n");
+    fprintf(stderr, "\t[-m pure reader cpu mask, default: 0x0 (no pure readers)]\n");
+}
+
+void tbb_parse_args(test_args unused, int argc, char** argv) {
+    int i = 0;
+
+    log2_ratio = 6;
+    reader_cpu_mask = 0x0;
+
+    while ((i = getopt(argc, argv, "hr:m:")) != -1)
+    {
+        switch (i) {
+          case 'r':
+            log2_ratio = strtoul(optarg, (char **) NULL, 10);
+            if (log2_ratio >= 64) {
+                fprintf(stderr, "tbb_spin_rw_mutex: -r can not be >= 64\n");
+                exit(1);
+            }
+            break;
+          case 'm':
+            if (!strncmp(optarg, "0x", 2))
+                reader_cpu_mask = strtoul(optarg, (char **) NULL, 16);
+            else
+                reader_cpu_mask = strtoul(optarg, (char **) NULL, 10);
+            if (errno == ERANGE) {
+                fprintf(stderr,
+                        "tbb_spin_rw_mutex: -m value unsuitable for 'unsigned long'\n");
+                exit (1);
+            }
+            break;
+          case 'h':
+            tbb_print_usage();
+            exit(0);
+          case '?':
+          default:
+            tbb_print_usage();
+            exit(3);
+        }
+        if (errno == EINVAL) {
```

Comment:
From errno manpage, errno may be any value if we don't check return value first:

> The value in errno is significant only when the return value of the call indicated an error (i.e., -1 from most system calls; -1 or NULL from most library functions); a function that succeeds is allowed to change errno.

Reply:
As with the previous case it seems impossible to tell whether or not a valid conversion occurred based on the return value of strtoul alone since "no conversion" returns 0 which is a valid option for both of the test-specific args.  Additionally the manpage suggests the EINVAL case is not portable anyway so maybe it's best to remove this and do our best to sanitize whatever the results from strtoul in the individual cases.